Tests/PackageLoading/InMemoryConventionTests.swift

```diff
@@ -0,0 +1,174 @@
+    let pkg: PackageDescription.Package
+    if let package = package {
+        pkg = package
+    } else {
+        pkg = Package(name: path.basename)
+    }
+    let manifest = Manifest(path: path.appending(component: Manifest.filename), url: "", package: pkg, products: [], version: nil)
+    let builder = PackageBuilder(manifest: manifest, path: path, fileSystem: fs)
+    return try builder.construct(includingTestModules: true)
+}
+
+extension PackageBuilderTester {
+    @discardableResult
+    init(_ path: AbsolutePath, in fs: FileSystem, package: PackageDescription.Package? = nil, result: (PackageBuilderTester) -> Void) {
+        do {
+            self = .package(try loadPackage(path, in: fs, package: package))
+        } catch {
+            self = .error(error)
+        }
+        result(self)
+    }
+}
+
+extension PackageModel.Package {
+    var allModules: [Module] {
+        return modules + testModules
+    }
+}
+
+enum PackageBuilderTester {
+    case package(PackageModel.Package)
+    case error(Swift.Error)
+
+    func checkError(_ str: String) {
+        switch self {
+        case .package(let package):
+            XCTFail("\(package) did not have any error")
+        case .error(let error):
+            XCTAssertEqual(String(error), str)
+        }
+    }
+
+    func checkModule(_ name: String, result: ((ModuleResult) -> Void)? = nil) {
+        guard case .package(let package) = self else {
+            return XCTFail("Error: \(self)")
+        }
+        guard let module = package.allModules.filter({$0.name == name}).first else {
+            return XCTFail("Module: \(name) not found")
+        }
+        result?(ModuleResult(module))
+    }
+
+    final class ModuleResult {
+        let module: Module
+        lazy var sources: Set<RelativePath> = { Set(self.module.sources.relativePaths) }()
+
+        init(_ module: Module) {
+            self.module = module
+        }
+
+        func check(c99name: String) {
+            XCTAssertEqual(module.c99name, c99name)
+        }
+
+        func check(type: ModuleType) {
+            XCTAssertEqual(module.type, type)
+        }
+
+        func check(isTest: Bool) {
+            XCTAssertEqual(module.isTest, isTest)
+        }
+
+        func checkSources(root: AbsolutePath) {
+            XCTAssertEqual(module.sources.root, root)
+        }
+
+        func checkSources(count: Int) {
+            XCTAssertEqual(module.sources.paths.count, count)
+        }
+
+        func checkSources(_ paths: RelativePath...) {
+            for path in paths {
+                XCTAssert(sources.contains(path), "\(path.asString) not found in module \(module.name)")
+            }
+        }
+    }
+}
+
+/// Tests for the handling of source layout conventions.
+class InMemoryConventionTests: XCTestCase {
+
+    func testMixedSources() throws {
+        var fs = InMemoryFileSystem()
+        try fs.createEmptyFiles("/home/MyPackage/Sources/main.swift",
```

Comment:
Could we write the tests rooted at `/` just to eliminate as much boilerplate as possible?